README
NAME
    Plack::Middleware::Debug::Mongo - Extend Plack::Middleware::Debug with
    MongoDB panels

VERSION
    version 0.02

SYNOPSIS
        # inside your psgi app
        use Plack::Builder;

        my $app = sub {[
            200,
            [ 'Content-Type' => 'text/html' ],
            [ '<html><body>OK</body></html>' ]
        ]};

        my $options = { host => 'mongodb://mongo.example.com:29111', db_name => 'sampledb' };

        builder {
            mount '/' => builder {
                enable 'Debug',
                    panels => [
                        [ 'Mongo::ServerStatus', connection => $options ],
                        [ 'Mongo::Database', connection => $options ],
                    ];
                $app;
            };
        };

DESCRIPTION
    This distribution extends Plack::Middleware::Debug with some MongoDB
    panels. At the moment, listed below panels are available.

PANELS
  Mongo::ServerStatus
    Display panel with MongoDB server status information. See
    Plack::Middleware::Debug::Mongo::ServerStatus for additional
    information.

  Mongo::Database
    Display panel with MongoDB database and its collections statistics. See
    Plack::Middleware::Debug::Mongo::Database for additional information.

BUGS
    Please report any bugs or feature requests through the web interface at
    <https://github.com/Wu-Wu/Plack-Middleware-Debug-Mongo/issues>

SEE ALSO
    Plack::Middleware::Debug::Mongo::ServerStatus

    Plack::Middleware::Debug::Mongo::Database

    Plack::Middleware::Debug

    MongoDB

    MongoDB Server Status Reference
    <http://docs.mongodb.org/manual/reference/server-status/>

    MongoDB Database Statistics Reference
    <http://docs.mongodb.org/manual/reference/database-statistics/>

    MongoDB Collection Statistics Reference
    <http://docs.mongodb.org/manual/reference/collection-statistics/>

AUTHOR
    Anton Gerasimov, <chim@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2013 by Anton Gerasimov

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Plack::Middleware::Debug::Mongo.

0.02  Mon Mar 4 12:01:07 2013
    - added hashwalk() export and updated *.t
    - added database statistics module and *.t
    - updated pods and readme

0.01  Wed Feb 27 15:33:04 2013
    - original version



Makefile.PL
use 5.008009;
use ExtUtils::MakeMaker;

my $git_repo = 'https://github.com/Wu-Wu/Plack-Middleware-Debug-Mongo';

my $meta_merge = {
    META_MERGE => {
        resources => {
            repository => $git_repo,
            bugtracker => $git_repo . '/issues',
        },
    }
};

my %options = (
    NAME          => 'Plack::Middleware::Debug::Mongo',
    VERSION_FROM  => 'lib/Plack/Middleware/Debug/Mongo.pm',
    AUTHOR        => 'Anton Gerasimov <chim@cpan.org>',
    ABSTRACT_FROM => 'lib/Plack/Middleware/Debug/Mongo.pm',
    LICENSE       => 'perl',
    PREREQ_PM     => {
        'strict'                   => 0,
        'warnings'                 => 0,
        'parent'                   => 0,
        'Plack'                    => 0,
        'Plack::Middleware::Debug' => 0,
        'MongoDB'                  => '0.502',
    },
    BUILD_REQUIRES => {
        'Test::More'       => '0.98',
        'Test::Pod'        => '1.22',
        'Plack'            => 0,
        'HTTP::Message'    => 0,
        'Test::MockObject' => 0,
        'boolean'          => 0,
    },
    MIN_PERL_VERSION => '5.8.9',
    test => {
        TESTS => 't/*.t',
    },
    clean => {
        FILES => '*.tar.gz',
    },
    $ExtUtils::MakeMaker::VERSION >= 6.50 ? (%$meta_merge) : (),
);

WriteMakefile(%options);


lib/Plack/Middleware/Debug/Mongo/Database.pm
package Plack::Middleware::Debug::Mongo::Database;

use strict;
use warnings;
use parent 'Plack::Middleware::Debug::Base';
use Plack::Util::Accessor qw/connection mongo_client/;
use MongoDB 0.502;
use Plack::Middleware::Debug::Mongo::ServerStatus 'hashwalk';

our $VERSION = '0.01';

sub prepare_app {
    my ($self) = @_;

    $self->connection->{db_name} = 'admin' unless exists $self->connection->{db_name};
    $self->mongo_client(MongoDB::MongoClient->new($self->connection));
}

sub run {
    my ($self, $env, $panel) = @_;

    $panel->title('Mongo::Database');
    $panel->nav_title($panel->title);

    my $dbh = $self->mongo_client->get_database($self->connection->{db_name});
    my @collections = sort grep { !/\$/ } $dbh->collection_names;

    $panel->nav_subtitle($self->connection->{db_name});
    my $info= {};

    my @elements = (
        'database: ' . $self->connection->{db_name},
    );

    # collection statistics
    foreach (@collections) {
        my $stats = {};
        hashwalk($dbh->run_command({collStats => $_}), $stats, undef);
        push @elements, 'collection: ' . $_;
        $info->{$elements[-1]} = $stats;
    }

    # database statistics
    $info->{$elements[0]} = $dbh->run_command({dbStats => 1});

    return sub {
        $panel->content($self->render_hash($info, [ @elements ]));
    };
}

1; # End of Plack::Middleware::Debug::Mongo::Database
__END__

=pod

=head1 NAME

Plack::Middleware::Debug::Mongo::Database - Mongo database debug panel for Plack::Middleware::Debug

=head1 VERSION

version 0.01

=head1 SYNOPSIS

    # inside your psgi app
    enable 'Debug',
        panels => [
            [ 'Mongo::Database', connection => $options ],
        ];

=head1 DESCRIPTION

Plack::Middleware::Debug::Mongo::Database extends Plack::Middleware::Debug by adding MongoDB database statistics debug panel
such as number of collections and indexes, average size of each document, total database and index size etc. This information
can be obtained by command I<db.stats()> in the mongo shell. See
L<MongoDB Database Statistics Reference|http://docs.mongodb.org/manual/reference/database-statistics/> for details.

Additionally collection's statistics displayed for each collection of the database. It includes number of documents and indexes,
total index and storage size, number of allocated data file regions, average size of an object in collection etc. More
information and detailed description of the result see
L<MongoDB Collection Statistics Reference|http://docs.mongodb.org/manual/reference/collection-statistics/>.

Sample output

    Database: awl_devel

    Key                     Value
    avgObjSize              643.318918918919
    collections             8
    dataSize                476056
    db                      awl_devel
    fileSize                201326592
    indexSize               81760
    indexes                 6
    nsSizeMB                16
    numExtents              11
    objects                 740
    ok                      1
    storageSize             585728

    Collection: inventory

    Key                     Value
    avgObjSize              171.651090342679
    count                   321
    flags                   1
    indexSizes._id_         24528
    lastExtentSize          49152
    nindexes                1
    ns                      awl_devel.inventory
    numExtents              2
    ok                      1
    paddingFactor           1
    size                    55100
    storageSize             61440
    totalIndexSize          24528

    ...

=head1 METHODS

=head2 prepare_app

See L<Plack::Middleware::Debug>

=head2 run

See L<Plack::Middleware::Debug>

=head2 connection

MongoDB connection options. Passed as HASH reference. Default server to connect is B<mongodb://localhost:27017>.
For additional information please consult L<MongoDB::MongoClient> page. Default database name is B<admin>. You should
change it to get fetch statistics of the desired database and its collections.

=head1 EXPORTED FUNCTIONS AND SUBROUTINES

Plack::Middleware::Debug::Mongo::Database doesn't export any functions and subroutines.

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Debug-Mongo/issues>

=head1 SEE ALSO

L<Plack::Middleware::Debug>

L<MongoDB::MongoClient>

L<MongoDB Database Statistics Reference|http://docs.mongodb.org/manual/reference/database-statistics/>

L<MongoDB Collection Statistics Reference|http://docs.mongodb.org/manual/reference/collection-statistics/>

=head1 AUTHOR

Anton Gerasimov, E<lt>chim@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Anton Gerasimov

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/Plack/Middleware/Debug/Mongo/ServerStatus.pm
package Plack::Middleware::Debug::Mongo::ServerStatus;

use strict;
use warnings;
use parent 'Plack::Middleware::Debug::Base';
use Plack::Util::Accessor qw/connection mongo_client/;
use MongoDB 0.502;
use Exporter 'import';

our $VERSION = '0.02';

our %EXPORT_TAGS = (
    'all' => [ qw(hashwalk) ]
);
our @EXPORT_OK   = (
    @{ $EXPORT_TAGS{'all'} }
);
our @EXPORT      = ();

sub prepare_app {
    my ($self) = @_;

    $self->connection->{db_name} = 'admin' unless exists $self->connection->{db_name};
    $self->mongo_client(MongoDB::MongoClient->new($self->connection));
}

sub run {
    my ($self, $env, $panel) = @_;

    $panel->title('Mongo::ServerStatus');
    $panel->nav_title($panel->title);

    my $status = $self->mongo_client->get_database($self->connection->{db_name})
                                    ->run_command({ serverStatus => 1 });
    $panel->nav_subtitle('Version: ' . $status->{version});
    my $info = {};

    hashwalk($status, $info, undef);

    return sub {
        $panel->content($self->render_hash($info));
    };
}

#
# flatten hash
sub hashwalk {
    my ($status, $info, $prefix) = @_;

    foreach my $key (keys %$status) {
        my $t_ref    = ref($status->{$key});
        my $t_prefix = $prefix ? join('.' => $prefix, $key) : $key;

        if ($t_ref eq 'HASH') {
            # next level
            hashwalk($status->{$key}, $info, $t_prefix);
        }
        elsif ($t_ref eq 'DateTime') {
            # convert DateTime
            $info->{$t_prefix} = $status->{$key}->datetime;
        }
        elsif ($t_ref eq 'boolean') {
            # convert boolean
            $info->{$t_prefix} = "$status->{$key}" ? 'true' : 'false';
        }
        else {
            $info->{$t_prefix} = $status->{$key};
        }
    }
}

1; # End of Plack::Middleware::Debug::Mongo::ServerStatus
__END__

=pod

=head1 NAME

Plack::Middleware::Debug::Mongo::ServerStatus - Server status debug panel for Plack::Middleware::Debug

=head1 VERSION

version 0.02

=head1 SYNOPSIS

    # inside your psgi app
    enable 'Debug',
        panels => [
            [ 'Mongo::ServerStatus', connection => $options ],
        ];

=head1 DESCRIPTION

Plack::Middleware::Debug::Mongo::ServerStatus extends Plack::Middleware::Debug by adding MongoDB server status debug panel.
Panel displays data which available through the I<db.serverStatus()> command issued in mongo CLI. Before displaying info some
tweaks were processed such as flatten several complex structures.

Sample output

    localTime                   2013-02-28T10:00:42
    mem.bits                    64
    mem.mapped                  1056
    mem.mappedWithJournal       2112
    mem.note                    not all mem info support on this platform
    mem.supported               false
    network.bytesIn             46182
    network.bytesOut            9697590320
    network.numRequests         587
    ok                          1

=head1 METHODS

=head2 prepare_app

See L<Plack::Middleware::Debug>

=head2 run

See L<Plack::Middleware::Debug>

=head2 connection

MongoDB connection options. Passed as HASH reference. Default server to connect is B<mongodb://localhost:27017>.
For additional information please consult L<MongoDB::MongoClient> page.

=head1 EXPORTED FUNCTIONS AND SUBROUTINES

Plack::Middleware::Debug::Mongo::ServerStatus doesn't export any subroutines by default. On request available the
following functions and subroutines.

=head2 hashwalk

Subroutine used to convert multidimensional hash references into simple hash reference. As well it converts DateTime and
boolean objects received from MongoDB into human readable format. The L<Plack::Middleware::Debug::Mongo::Database> uses
this subroutine.

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Debug-Mongo/issues>

=head1 SEE ALSO

L<Plack::Middleware::Debug::Mongo::Database>

L<Plack::Middleware::Debug>

L<MongoDB::MongoClient>

L<DateTime>

L<boolean>

L<MongoDB Server Status Reference|http://docs.mongodb.org/manual/reference/server-status/>

=head1 AUTHOR

Anton Gerasimov, E<lt>chim@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Anton Gerasimov

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/Plack/Middleware/Debug/Mongo.pm
package Plack::Middleware::Debug::Mongo;

use strict;
use warnings;

our $VERSION = '0.02';

1; # End of Plack::Middleware::Debug::Mongo
__END__

=head1 NAME

Plack::Middleware::Debug::Mongo - Extend Plack::Middleware::Debug with MongoDB panels

=head1 VERSION

version 0.02

=head1 SYNOPSIS

    # inside your psgi app
    use Plack::Builder;

    my $app = sub {[
        200,
        [ 'Content-Type' => 'text/html' ],
        [ '<html><body>OK</body></html>' ]
    ]};

    my $options = { host => 'mongodb://mongo.example.com:29111', db_name => 'sampledb' };

    builder {
        mount '/' => builder {
            enable 'Debug',
                panels => [
                    [ 'Mongo::ServerStatus', connection => $options ],
                    [ 'Mongo::Database', connection => $options ],
                ];
            $app;
        };
    };

=head1 DESCRIPTION

This distribution extends Plack::Middleware::Debug with some MongoDB panels. At the moment, listed below panels are
available.

=head1 PANELS

=head2 Mongo::ServerStatus

Display panel with MongoDB server status information.
See L<Plack::Middleware::Debug::Mongo::ServerStatus> for additional information.

=head2 Mongo::Database

Display panel with MongoDB database and its collections statistics.
See L<Plack::Middleware::Debug::Mongo::Database> for additional information.

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Debug-Mongo/issues>

=head1 SEE ALSO

L<Plack::Middleware::Debug::Mongo::ServerStatus>

L<Plack::Middleware::Debug::Mongo::Database>

L<Plack::Middleware::Debug>

L<MongoDB>

L<MongoDB Server Status Reference|http://docs.mongodb.org/manual/reference/server-status/>

L<MongoDB Database Statistics Reference|http://docs.mongodb.org/manual/reference/database-statistics/>

L<MongoDB Collection Statistics Reference|http://docs.mongodb.org/manual/reference/collection-statistics/>

=head1 AUTHOR

Anton Gerasimov, E<lt>chim@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Anton Gerasimov

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
